### DISCUSSION TITLE
Issues with using await in conjunction with terminal.sendText in VSCode extensions

### DISCUSSION THREAD

**Discussion by ji-podhead** [microsoft/vscode-discussions#1091] [Extension Development QnA] [answered]
hi and greetings from Germany <3

During the development of a Visual Studio Code extension that involves executing commands in a terminal window, I've encountered unexpected behavior. I'm attempting to monitor the execution of commands in a terminal window by using await in conjunction with a custom waitForCommandCompletion function that waits for the success of a command before proceeding to the next step.

The issue arises when the await expression is resolved immediately because the command has already been sent to the terminal but has not yet been executed. I think the terminal operates as a so-called "pseudoterminal," which queues commands to be executed sequentially, but await expects an immediate resolution, leading to unexpected behavior.

I suspect that the mutability of the terminal (the ability to change the state of the terminal during command execution) plays a role, as it determines how and when data can be modified. When using await to wait for the completion of an asynchronous operation, one would expect the state being awaited to remain unchanged while waiting. In this case, the problem is that the state of the terminal can change after the command has been sent but before it has been executed.

This behavior makes it difficult to manage sequential commands in a terminal window, especially when the execution of the next command depends on the completion of the previous one. A possible solution would be to redundantly write the code by executing the commands directly in the extension, rather than sending them to the terminal. Another possibility would be to pass the terminal object to the method that executes the commands, but this would compromise the reusability of the method and would not be ideal.

One specific method that is causing issues in my code is create_protobuffs. This method is called after a series of commands have been sent to the terminal, and it is expected to execute only after these commands have completed. However, due to the asynchronous nature of the terminal and the use of await, this method is being called before the commands have actually finished executing, leading to errors.
i have looked up a bunch of similar posts like this one here: https://stackoverflow.com/questions/55943439/how-to-wait-until-vscode-windows-terminal-operation-is-over
and i tried to send a msg and await for it. that doesnt work so i thought of creating a finish.txt that indicates that the process is finished, but that all seem to be a hacky workaround.

I would like to know if there is a recommended method to work around this issue or if there are plans to change or improve this behavior in future versions of VSCode.

Version Information:

Operating System: Linux
VSCode Version: {1.87.0
019f4d1419fbc8219a181fab7892ebccf7ee29a2
x64
}
Expected Behavior: A way to monitor the execution of commands in a terminal window and ensure that commands are executed sequentially and in the correct order without await being resolved immediately.

Current Behavior: await expressions are resolved immediately because the commands have already been sent to the terminal but have not yet been executed.

```
async function waitForCommandCompletion(terminal, successMessage, timeout = 300000) {
    return new Promise((resolve, reject) => {
        const startTime = Date.now();
        const interval = setInterval(() => {
            // Überprüfen Sie, ob der Erfolgstext im Terminal-Ausgabepfad erscheint
            // Dies ist nur ein Platzhalter. Sie müssen die Logik anpassen, um den spezifischen Erfolgstext zu überprüfen.
            const output = terminal.output; // Angenommen, `terminal.output` gibt den aktuellen Ausgabetext des Terminals zurück
            if (output.includes(successMessage)) {
                clearInterval(interval);
                resolve();
            } else if (Date.now() - startTime > timeout) {
                clearInterval(interval);
                reject(new Error('Timeout while waiting for command completion'));
            }
        }, 10000); // Überprüfen Sie alle 10000 Millisekunden (1 Sekunde)
    });
}

async function initialize_project() {
    return new Promise((resolve, reject) => {
      
        cp.exec('air -v', async (error, stdout, stderr) => {
            if (error) {
                // Air ist nicht installiert, installieren Sie es
                vscode.window.showInformationMessage('Air ist nicht installiert. Installiere Air jetzt...');
                const command = 'curl -sSfL https://raw.githubusercontent.com/cosmtrek/air/master/install.sh | sh -s -- -b $(go env GOPATH)/bin';
                var terminal = vscode.window.createTerminal({
                    name: "Air Installation",
                    cwd: vscode.workspace.workspaceFolders[0].uri.fsPath
                });
                terminal.sendText(<command>);                    
                // a bunch of additional commands 
                terminal.sendText('echo "init ji_ui finished successfully"');                                                   //    <------------------ 

                try {
                    await waitForCommandCompletion(terminal, 'init ji_ui finished successfully');            //    <------------------ 
                   
                    create_protobuffs();                                                                                                               //    <------------------ 
                    terminal.dispose()
                    resolve()
                } catch (error) {
                    vscode.window.showErrorMessage('Fehler bei der Projektinitialisierung: ' + error);
                }
            
                resolve()
            } else {
                // Air ist installiert, fahren Sie mit der Konfiguration fort
                vscode.window.showInformationMessage('Air ist bereits installiert.');
                reject();
            }
        });
    })
}
```
Help is much appreciated!
Thanks in advance,
best regards:
leo/ji/podhead

**Comment by ji-podhead** [comment]
so the hack where i wait for a certain file to get  created works, but my waitForCommandCompletion() timed out and i cant tell how long it takes for the initialize_project method to finish because that depends on the users resources, internet connection etc...
so i asked phind about the pseudoterminal api and if that could be used to solve my problem, but it seems that this is more for the extension development itself rather than for the code window that actually uses the extension

**Reply by ji-podhead** [reply]
added a [feature request for this issue](https://github.com/microsoft/vscode/issues/207158)

**Comment by r3m0t** [comment]
Hi @ji-soft

Terminal is just a dumb display of characters and VS Code doesn't know if your command completed or not.

The exception is with shell integration - https://code.visualstudio.com/docs/terminal/shell-integration - but you should not assume this works with every machine or every configuration.

The best way to get output is with the `child_process` module, or if the user needs to see the output, use `executeTask`.

**Reply by ji-podhead** [reply]
ahh allright a cp can be promisified. thanks for that.
But how can i use the vscode.window.createTerminal inside that cp without creating another script that i pass to exec as a string?
and i still dont know if the cp promise will resolve directly, or after all that commands are completed in the workspace that runs the extension. 
so the other option would be to use a bashscript instead of passing in the commands one by one to the terminal with executeTask, but this requires to add the tasks to the tasks.json of the extension workspace?
or do i need to add the task to the workspace that uses the extension?

**Reply by r3m0t** [reply]
You can call `executeTask` and define a `ShellExecution` or `ProcessExecution`. The user won't need to edit tasks.json.

I think writing a bash script to a temporary folder and executing it is a good idea.

**Reply by ji-podhead** [reply]
thanks i will try this now:

- copy the bashscript over to the vscode.workspace.workspaceFolders[0].uri.fsPath;
- promisify the childprocess and await exec(bash my_bashscript)
- wait for  child.on('close', or just await the cp

so i guess i will try something like this:

```
const cp = require('child_process');
const util = require('util');
const exec = util.promisify(cp.exec);
async function runScript() {
    const workspacePath = vscode.workspace.workspaceFolders[0].uri.fsPath;
    const scriptPath = path.join(os.tmpdir(), 'myScript.sh');
    const scriptContent = `#!/bin/bash
echo "Dies ist ein Bash-Skript."
# Weitere Befehle...
`;
    fs.writeFileSync(scriptPath, scriptContent);
    fs.chmodSync(scriptPath, 0o7555);
    const child = cp.spawn('bash', [scriptPath], { cwd: workspacePath });
    child.stdout.on('data', (data) => {
        console.log(`stdout: ${data}`);
    });
    child.stderr.on('data', (data) => {
        console.error(`stderr: ${data}`);
    });
    child.on('close', (code) => {
        console.log(`child process exited with code ${code}`);
    });
    try {
        await new Promise((resolve, reject) => {
            child.on('error', reject);
            child.on('exit', (code) => {
                if (code === 0) {
                    resolve();
                } else {
                    reject(new Error(`Process exited with code ${code}`));
                }
            });
        });
        console.log('Projektinitialisierung abgeschlossen.');
        after_bash_execution();
    } catch (error) {
        console.error('Fehler bei der Projektinitialisierung:', error);
    }
}
function after_bash_execution() {
    // Ihre Logik hier
    console.log('Protobufs generieren...');
}


runScript();

``` 

another option would be to pass an array of strings to a function that creates a cp for every command in the array, that prevents the creation of a bash script that i need to copy to temp directory in the extension-user-workspace.
i also found that you can pass the output to another terminal so the user can tell what actually happens :  

```
 child.stdout.on('data', (data) => {
                   var terminal = vscode.window.createTerminal({
                    name: "log-terminal",
                    cwd: vscode.workspace.workspaceFolders[0].uri.fsPath
                });
                terminal.sendText(echo data);
    });
```

but this like a lot of code/work for such simple thing, so i guess i create a wrapper class using my second approach.
i think this will solve it!!! so huge  thanks to you @r3m0t for getting me on the right path!!! 
Ill close that issue now!
enjoy your weekend ;)

**Reply by Tyriar** [reply]
Good answer from @r3m0t - luckily I'm looking at the shell integration extension API again this month which will accomplish exactly what you're after: https://github.com/microsoft/vscode/issues/145234.

Note that using a child process comes with the downside of not being a pty and therefore not being interactive/colored/etc.

**Comment by ji-podhead** [comment]
>... so i guess i create a wrapper class using my second approach. i think this will solve it!!!

This is my Wrapper class using childprocess. You can pass a single command, or an array of commands.

but as said above its  not interactive, colored etc and does feel a bit weird. and it doesnt work for git pull somehow. so i dont gonna use it  :( :
```
const cp = require('child_process');
const util = require('util');
const path = require('path');
const vscode = require('vscode');

// Promisify cp.spawn
const spawn = util.promisify(cp.spawn);

class CommandExecutor {
    constructor() {
        this.terminal = null;
    }

    initialize() {
        if (!this.terminal) {
            this.terminal = vscode.window.createTerminal({
                name: "Command Execution",
                cwd: vscode.workspace.workspaceFolders[0].uri.fsPath
            });
            this.terminal.show();
        }
    }

    dispose() {
        if (this.terminal) {
            this.terminal.dispose();
            this.terminal = null;
        }
    }

    async runCommands(commands) {
        this.initialize();
        for (const command of commands) {
            await this.executeCommand(command);
        }
        this.dispose();
    }

    async executeCommand(command) {
        this.terminal.sendText(command);

        return new Promise((resolve, reject) => {
            const child = cp.spawn('bash', ['-c', command], {
                cwd: vscode.workspace.workspaceFolders[0].uri.fsPath
            });

            let stdout = '';
            let stderr = '';

            child.stdout.on('data', (data) => {
                stdout += data;
                this.terminal.sendText(data.toString());
            });

            child.stderr.on('data', (data) => {
                stderr += data;
                this.terminal.sendText(data.toString());
            });

            child.on('error', (error) => {
                reject(error);
            });

            child.on('exit', (code) => {
                if (code === 0) {
                    resolve({ stdout, stderr });
                } else {
                    reject(new Error(`Process exited with code ${code}`));
                }
            });
        });
    }
}

```
# Run

```
const commandExecutor = new CommandExecutor();
const commands = ['sleep 10','echo "Hello, World!"', 'sleep 2 && ls -la,'sleep 10','echo "Hello, World!"', 'sleep 2 && ls -la'];
commandExecutor.runCommands(commands).then(() => {
    console.log('All commands executed successfully.');
}).catch(error => {
    console.error('Error executing commands:', error);
});
```

**Reply by r3m0t** [reply]
If you're sending the output using `sendText` doesn't that mean you are submitting it as a command to a shell? Could be dangerous. You can use `createOutputChannel` instead to display the output to the user.

**Reply by ji-podhead** [reply]
biggest problem here though is that i cant use git pull commands with this approach.
i think they fail due to async somehow.
this was failing for me:
```
    const commands2=[
        'curl -fsSL https://bun.sh/install | bash',
        'mkdir frontend',
        'bun create react-app frontend',
        'cd frontend',
        'bun install google-protobuf',
        'bun install @improbable-eng/grpc-web',
        "bun install ts-protoc-gen",
        'cd src',
        'git clone https://github.com/ji-soft/ji_ui_react_example',
        'cp -rf ji_ui_react_example/* .',
        'chmod -R u+w ji_ui_react_example',
        'rm -rf ji_ui_react_example',
    ]
 
    
    commandExecutor.runCommands(commands2).then(() => {
```
    
but the same commands in the same order works just fine using window.terminal directly. super weird

**Reply by ji-podhead** [reply]
so  ill wait till there is a proper implementation, but i guess its almost ready: see [proposal](https://github.com/microsoft/vscode/issues/145234#issuecomment-1986252477)

**Comment by abrilurena** [comment]
Hi, @ji-podhead! I hope you're doing great. Just wondering, do you still need help with this?

**Reply by ji-podhead** [reply, marked as answer]
Hey, thank you for the reply. I'm currently waiting for new terminal features to become available. I'm currently working on something else anyway. I can close this discussion for now, I intentionally wanted to close it with a link to the new changes.

**Reply by abrilurena** [reply]
Hi, @ji-podhead! hope you're doing great. Awesome, thank you! then, i'll leave it open :) This item appears to be answered, so marking it as such.
